File: 백준/2917_늑대사냥꾼/Main.java
package algorithm;

import java.io.*;
import java.util.*;

class Solution {

    private BufferedReader reader;

    private BufferedWriter writer;

    private int N, M;

    private char[][] grid;

    private int[] dr = {0, 1, 0, -1};

    private int[] dc = {1, 0, -1, 0};

    private static final char EMPTY = '.';

    private static final char TREE = '+';

    private static final char SHELTER = 'J';

    private static final char WOLF = 'V';

    private int[] wolfPos;

    private int[][] distFromTree;

    private boolean[][] visited;

    private List<int[]> trees;

    public Solution(BufferedReader reader, BufferedWriter writer) throws IOException {
        this.reader = reader;
        this.writer = writer;

        String[] tokens = reader.readLine().split(" ");
        N = Integer.parseInt(tokens[0]);
        M = Integer.parseInt(tokens[1]);
        grid = new char[N][M];
        wolfPos = new int[2];
        trees = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            String line = reader.readLine();
            for (int j = 0; j < M; j++) {
                grid[i][j] = line.charAt(j);

                if ( grid[i][j] == WOLF ) {
                    wolfPos[0] = i; // row
                    wolfPos[1] = j; // column
                } else if ( grid[i][j] == TREE ) {
                    trees.add(new int[]{i, j});
                }
            }
        }

        visited = new boolean[N][M];
        distFromTree = new int[N][M];
    }

    private boolean isInRange(int r, int c) {
        return r >= 0 && r < N && c >= 0 && c < M;
    }

    private void calculateDistanceFromTrees() {
        Queue<int[]> q = new ArrayDeque();

        for (int[] tree : trees) {
            q.add(new int[]{tree[0], tree[1], 0});
            visited[tree[0]][tree[1]] = true;
        }

        while ( !q.isEmpty() ) {
            int[] cur = q.poll();

            for ( int i = 0 ; i < 4 ; ++i ) {
                int nr = cur[0] + dr[i];
                int nc = cur[1] + dc[i];

                if ( isInRange(nr, nc) && !visited[nr][nc] ) {
                    visited[nr][nc] = true;
                    distFromTree[nr][nc] = cur[2] + 1;
                    q.add(new int[]{nr, nc, cur[2] + 1});
                }
            }
        }
    }

    private void printDistanceFromTrees() {
        System.out.println("Distance from trees:");
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                System.out.print(distFromTree[i][j] + " ");
            }
            System.out.println();
        }
    }

    private void simulate() throws IOException { 
        calculateDistanceFromTrees();

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[2] - a[2]);

        for ( boolean[] row : visited ) {
            Arrays.fill(row, false);
        }

        // printDistanceFromTrees();
        pq.add(new int[]{wolfPos[0], wolfPos[1], distFromTree[wolfPos[0]][wolfPos[1]]});
        visited[wolfPos[0]][wolfPos[1]] = true;
        int answer = Integer.MAX_VALUE;

        while ( !pq.isEmpty() ) {
            int[] cur = pq.poll();
            // System.out.println("Current position: " + cur[0] + ", " + cur[1] + ", Distance: " + cur[2]);

            answer = Math.min(answer, cur[2]);

            if ( grid[cur[0]][cur[1]] == SHELTER ) {
                break;
            }

            for ( int i = 0 ; i < 4 ; ++i ) {
                int nr = cur[0] + dr[i];
                int nc = cur[1] + dc[i];

                if ( isInRange(nr, nc) && !visited[nr][nc] ) {
                    visited[nr][nc] = true;
                    // System.out.println("Adding to PQ: " + nr + ", " + nc + ", " + distFromTree[nr][nc]);
                    pq.add(new int[]{nr, nc, distFromTree[nr][nc]});
                }
            }
        }
        writer.write(String.valueOf(answer) + "\n");
    }

    public void run() throws IOException {
        simulate();
        writer.flush();
    }
}

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        new Solution(reader, writer).run();
        reader.close();
        writer.close();
    }
}

File: 백준/2917_늑대사냥꾼/main.kt
package kr.dohoon_kim.algorithm.kotlin

import java.io.*;
import java.util.*;
import kotlin.io.path.name


class Solution(
    private val reader: BufferedReader,
    private val writer: BufferedWriter
) {

    private final val WOLF = 'V'

    private final val TREE = '+'

    private final val EMPTY = '.'

    private final val SHELTER = 'J'

    private var N = 0

    private var M = 0

    private var grid: Array<Array<Char>>

    private var visited: Array<Array<Boolean>>
    
    private var distFromTrees: Array<IntArray>

    private val trees = mutableListOf<IntArray>()

    private var sr = 0

    private var sc = 0

    private val dr = intArrayOf(-1, 0, 1, 0)

    private val dc = intArrayOf(0, 1, 0, -1)

    init {
        var line : List<String> = this.reader.readLine().split(" ")
        N = line[0].toInt()
        M = line[1].toInt()

        grid = Array(N) { Array(M) { ' ' } }

        for (i in 0 until N) {
            val l = this.reader.readLine()
            for (j in 0 until M) {
                grid[i][j] = l[j]

                if (grid[i][j] == TREE) {
                    trees.add(intArrayOf(i, j))
                }

                if (grid[i][j] == WOLF) {
                    sr = i
                    sc = j
                }
            }
        } 

        visited = Array(N) { Array(M) { false } }
        distFromTrees = Array(N) { IntArray(M) { 0 } }
    }

    private fun Int.isInRange(size: Int) = this in 0 until N

    private fun Pair<Int, Int>.isInGrid(): Boolean {
        val (r, c) = this
        return r.isInRange(N) && c.isInRange(M)
    }

    private fun calcDistFromTrees() {
        val q: Queue<IntArray> = ArrayDeque<IntArray>()

        for (tree in trees) {
            val r = tree[0]
            val c = tree[1]
            q.add(intArrayOf(r, c, 0))
            visited[r][c] = true
            distFromTrees[r][c] = 0
        }

        while (!q.isEmpty()) {
            val cur = q.poll()
            val r = cur[0]
            val c = cur[1]
            val d = cur[2]

            for (i in 0 until 4) {
                val nr = r + dr[i]
                val nc = c + dc[i]

                if (Pair(nr, nc).isInGrid() && !visited[nr][nc]) {
                    visited[nr][nc] = true
                    distFromTrees[nr][nc] = d + 1
                    q.add(intArrayOf(nr, nc, d + 1))
                }
            }
        }
    }

    private fun printDist() {
        println("Distance from trees:")
        for (i in 0 until N) {
            for (j in 0 until M) {
                writer.write("${distFromTrees[i][j]} ")
            }
            writer.write("\n")
        }
    }

    private fun calcPath(): Int {
        var answer = Int.MAX_VALUE
        val pq: PriorityQueue<IntArray> = PriorityQueue { a, b -> b[2] - a[2] }
        for ( row in visited ) {
            Arrays.fill(row, false)
        }
        // printDist()

        pq.add(intArrayOf(sr, sc, distFromTrees[sr][sc]))

        while (!pq.isEmpty()) {
            val cur = pq.poll()
            val r = cur[0]
            val c = cur[1]
            val d = cur[2]

            answer = Math.min(answer, d)
            if ( grid[r][c] == SHELTER ) {
                break
            }

            for (i in 0 until 4) {
                val nr = r + dr[i]
                val nc = c + dc[i]

                if (Pair(nr, nc).isInGrid() && !visited[nr][nc]) {
                    visited[nr][nc] = true
                    pq.add(intArrayOf(nr, nc, distFromTrees[nr][nc]))
                }
            }
        }

        return answer
    }

    fun run() {
        calcDistFromTrees()
        writer.write("${calcPath()}\n")
        writer.flush();
    }
}


fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.out))
    Solution(reader, writer).run()
}
